### ISSUE TITLE
How to provide arguments to custom class?

### ISSUE THREAD

**Issue by SamuelMarks** [shx2/apegears#1]
Attempt:
```python
@classmethod
def from_string(cls, s):
    return cls(**loads(s))

@classmethod
def __argparse__(cls):
    return dict(from_string=cls.from_string)
```

But this just puts it all on the first argument, like so:
```python
$ ipython3
Python 3.8.6 (default, Oct  8 2020, 14:06:32) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.0.0.dev -- An enhanced Interactive Python. Type '?' for help.

In [1]: from apegears import ArgumentParser, register_spec

In [2]: from ml_params_tensorflow.ml_params.config import TrainConfig

In [3]: from json import loads

In [4]: parser = ArgumentParser()

In [5]: parser.add_argument('-t', type=TrainConfig)
Out[5]: _StoreAction(option_strings=['-t'], dest='t', nargs=None, const=None, default=None, type=<class 'ml_params_tensorflow.ml_params.config.TrainConfig'>, choices=None, help=None, metavar=None)

In [6]: parser.parse_args(['-t', '{"epochs": 1, "loss": "CategoricalCrossentropy", "optimizer": "Adam", "callbacks": ["TensorBoard"], "metrics": [], "m
   ...: etric_emit_freq": None, "save_directory": None, "output_type": "infer", "validation_split": 0.2, "batch_size": 128, "kwargs": None, "return_typ
   ...: e": None}'])
Out[6]: Namespace(t=TrainConfig(epochs='{"epochs": 1, "loss": "CategoricalCrossentropy", "optimizer": "Adam", "callbacks": ["TensorBoard"], "metrics": [], "metric_emit_freq": None, "save_directory": None, "output_type": "infer", "validation_split": 0.2, "batch_size": 128, "kwargs": None, "return_type": None}', loss=None, optimizer=None, callbacks=None, metrics=None, metric_emit_freq=None, save_directory=None, output_type='infer', validation_split=0.1, batch_size=128, kwargs=None, return_type=<_collections._tuplegetter object at 0x106af1970>))
```

I'm sure I'm doing something wrong. I've also tried without `classmethod` and just manipulating the `class` symbol that I import in…

What's the right way to pass many arguments to the given `class` type? - Also is there a way to give per argument `--help` text and errors?

Happy to help implement. https://github.com/SamuelMarks/doctrans is what I'm working on.

Thanks for making apegears

**Comment by shx2** [comment]
Hi Samuel,

There are two problems in your code.

Firstly, argument-specs only work with `add_optional()`, `add_positional()`, etc., but not when using `add_argument()` directly.  In general, with apegears you should always prefer using one of the specialized `add_xxx()` methods, over the general-purpose `add_argument()` method.

Secondly, like the docs say, `__argparse__` should be a _class property_, not a _class method_.  You can use [this useful classproperty decorator](https://stackoverflow.com/a/17007928/2096752).

The following works for me:

    class TrainConfig:
    
        def __init__(self, **kw):
            self.kw = kw
    
        @classmethod
        def from_string(cls, s):
            return cls(**loads(s))
        
        @classproperty
        def __argparse__(cls):
            return dict(from_string=cls.from_string)
    
    from apegears import ArgumentParser, register_spec
    from json import loads
    parser = ArgumentParser()
    parser.add_optional('t', type=TrainConfig)
    parser.parse_args(['-t', '{"epochs": 1, "loss": "CategoricalCrossentropy", "optimizer": "Adam"}']).t.kw




Thank you for using and contributing to apegears.

**Comment by shx2** [comment]
> Also is there a way to give per argument --help text and errors?

Yes, default values for `help` and `metavar` can be specified in the arg-spec.
See [here](https://github.com/shx2/apegears/blob/master/apegears/spec.py#L15).

**Comment by shx2** [comment]
If you'd like to add to apegears a "json" arg spec, that would be cool.

**Comment by SamuelMarks** [comment]
Hmm… yeah happy to help out with the JSON arg spec, although TBH there has to be a nicer syntax. I'll probably just end up creating a marshall type that can abstract away anything that implements `loads`.

Anywhere, here's where I got up to with your aforementioned advice:

```python
In [1]: from apegears import ArgumentParser

In [2]: from ml_params_tensorflow.ml_params.config import TrainConfig

In [3]: parser = ArgumentParser()

In [4]: parser.add_optional('-t', type=TrainConfig)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-af7c2d0932ce> in <module>
----> 1 parser.add_optional('-t', type=TrainConfig)

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/parser.py in ArgumentParser.add_optional(self, *flags, **kwargs)
    168 def add_optional(self, *flags, **kwargs):
    169     """
    170     Add an *optional* parameter.  This calls ``add_argument`` with appropriate values.
    171 
   (...)
    178         is non-positional, but is not necessarilly *optional*, i.e. you can set required=True.
    179     """
--> 180     flags, kwargs = self._use_spec(*flags, is_positional=False, **kwargs)
    181     if not flags:
    182         raise ValueError('name not supplied for optional')

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/parser.py in ArgumentParser._use_spec(self, is_positional, *args, **kwargs)
    332 type = kwargs.get('type', None)
    333 if type is not None:
--> 334     spec = self._spec_from_type(type)
    335     if spec is not None:
    336         args, kwargs = self._apply_spec(spec, *args, is_positional=is_positional, **kwargs)

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/parser.py in ArgumentParser._spec_from_type(self, type)
    371 if type is None:
    372     return None
--> 373 spec = _find_spec(type)
    374 if spec is not None:
    375     return spec

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/spec.py in find_spec(cls)
     50 spec = getattr(cls, '__argparse__', None)
     51 if spec is not None:
---> 52     return to_spec(spec)
     53 # we can generate sensible specs for some types on the fly:
     54 spec = gen_type_spec(cls)

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/spec.py in to_spec(spec)
     74     attr_dict = {k: v for k, v in spec.__dict__.items() if not k.startswith('_')}
     75 else:
---> 76     attr_dict = dict(spec)
     77 return ArgParseSpec(**attr_dict)

TypeError: 'classproperty' object is not iterable
```

With these added:
```python
@classmethod
def from_string(cls, s):
    return cls(**loads(s))

@classproperty
def __argparse__(cls):
    return dict(from_string=cls.from_string)
```

Tried with decorating my [`dataclasses.dataclass`](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) and [`runtime_type_checker.check_types`](https://pypi.org/project/runtime-type-checker). Also tried with [`attr.s(auto_attribs=True)`](https://pypi.org/project/attrs)

Basically I'm trying to make `apegears` with my `class`—and [`doctrans`](https://github.com/SamuelMarks/doctrans)—behave like [`structopt`](https://github.com/TeXitoi/structopt) [and be just as concise]

**Comment by shx2** [comment]
what's the output of this: `print(TrainConfig.__argparse__)`?

**Comment by SamuelMarks** [comment]
[`TrainConfig`](https://github.com/SamuelMarks/ml-params-tensorflow/blob/1d48502/ml_params_tensorflow/ml_params/config.py#L15) with those aforementioned edits, gets the following output from your debug:
```
<ml_params_tensorflow.ml_params.config.classproperty object at 0x139ccc070>
```

**Comment by shx2** [comment]
I'm not sure what you're doing differently, because you didn't show the full definition of your class, but this works for me:

    class classproperty(object):
         def __init__(self, f):
             self.f = classmethod(f)
         def __get__(self, *a):
             return self.f.__get__(*a)()
      
     class TrainConfig:
         @classmethod
         def from_string(cls, s):
             return cls(**loads(s))
         @classproperty
         def __argparse__(cls):
             return dict(from_string=cls.from_string)
     
     print(TrainConfig.__argparse__)
     
    {'from_string': <bound method TrainConfig.from_string of <class '__main__.TrainConfig'>>}

**Comment by SamuelMarks** [comment]
I linked it earlier, but here's an inline minimal representative example `class` definition:
```python
from typing import Literal
from dataclasses import dataclass
from runtime_type_checker import check_types


class classproperty(object):
    def __init__(self, f):
        self.f = classmethod(f)

    def __get__(self, *a):
        return self.f.__get__(*a)()


def run_typed(f):
    return check_types(dataclass(f))


@run_typed
class TrainConf(object):
    optimizer: Literal[
        "Adadelta", "Adagrad", "Adam", "Adamax", "Ftrl", "Nadam", "RMSprop"
    ] = None

    @classmethod
    def from_string(cls, s):
        return cls(**loads(s))

    @classproperty
    def __argparse__(cls):
        return dict(from_string=cls.from_string)
```


```python
Python 3.8.6 (default, Oct  8 2020, 14:06:32) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.0.0.dev -- An enhanced Interactive Python. Type '?' for help.

In [1]: from ml_params_tensorflow.ml_params.config import TrainConf

In [2]: from apegears import ArgumentParser

In [3]: parser = ArgumentParser()

In [4]: parser.add_optional('-t', type=TrainConf)
spec: <ml_params_tensorflow.ml_params.config.classproperty object at 0x1492353d0> ;
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-26293bffa348> in <module>
----> 1 parser.add_optional('-t', type=TrainConf)

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/parser.py in ArgumentParser.add_optional(self, *flags, **kwargs)
    168 def add_optional(self, *flags, **kwargs):
    169     """
    170     Add an *optional* parameter.  This calls ``add_argument`` with appropriate values.
    171 
   (...)
    178         is non-positional, but is not necessarilly *optional*, i.e. you can set required=True.
    179     """
--> 180     flags, kwargs = self._use_spec(*flags, is_positional=False, **kwargs)
    181     if not flags:
    182         raise ValueError('name not supplied for optional')

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/parser.py in ArgumentParser._use_spec(self, is_positional, *args, **kwargs)
    332 type = kwargs.get('type', None)
    333 if type is not None:
--> 334     spec = self._spec_from_type(type)
    335     if spec is not None:
    336         args, kwargs = self._apply_spec(spec, *args, is_positional=is_positional, **kwargs)

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/parser.py in ArgumentParser._spec_from_type(self, type)
    371 if type is None:
    372     return None
--> 373 spec = _find_spec(type)
    374 if spec is not None:
    375     return spec

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/spec.py in find_spec(cls)
     50 spec = getattr(cls, '__argparse__', None)
     51 if spec is not None:
---> 52     return to_spec(spec)
     53 # we can generate sensible specs for some types on the fly:
     54 spec = gen_type_spec(cls)

~/repos/.venvs_py3/ml-env/lib/python3.8/site-packages/apegears/spec.py in to_spec(spec)
     75 else:
---> 76     attr_dict = dict(spec)
     77 return ArgParseSpec(**attr_dict)

TypeError: 'classproperty' object is not iterable
```

**Comment by shx2** [comment]
Seems like `run_typed` preventes `classproperty` from working properly.   I don't know why.

There are alternative ways to enable the spec and make this work.

Option 2: using `__argparse__` attribute, setting it explicitly:

    class TrainConfig:
        ...
    TrainConfig.__argparse__ = dict(from_string=TrainConfig.from_string)

Option 3: using `register_spec`:

    register_spec(TrainConfig, dict(from_string=TrainConfig.from_string))

**Comment by SamuelMarks** [comment]
Hmm, okay `check_types` seems to be causing the issue… so maybe I'll try some other runtime type checking library?:
```python
from typing import Literal

from apegears import ArgumentParser, register_spec
from dataclasses import dataclass
from runtime_type_checker import check_types


@check_types
class TrainConf(object):
    optimizer: Literal[
        "Adadelta", "Adagrad", "Adam", "Adamax", "Ftrl", "Nadam", "RMSprop"
    ] = None

    @classmethod
    def from_string(cls, s):
        return cls(**loads(s))

register_spec(TrainConf, dict(from_string=TrainConf.from_string))

ArgumentParser().add_optional('-t', type=TrainConf)
```

```
/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/argparse.py in _ActionsContainer.add_argument(self, *args, **kwargs)
   1371 type_func = self._registry_get('type', action.type, action.type)
   1372 if not callable(type_func):
-> 1373     raise ValueError('%r is not callable' % (type_func,))
   1375 if type_func is FileType:
   1376     raise ValueError('%r is a FileType class object, instance of it'
   1377                      ' must be passed' % (type_func,))

ValueError: <classmethod object at 0x111792910> is not callable
```

**Comment by shx2** [comment]
Yep. This isn't an apegears issue.
closing.
